src/ccsr/tls.rs

```diff
@@ -9,49 +9,53 @@
 
 //! TLS certificates and identities.
 
+use openssl::pkcs12::Pkcs12;
+use openssl::pkey::PKey;
+use openssl::stack::Stack;
+use openssl::x509::X509;
 use serde::{Deserialize, Serialize};
 
-// Encodes the type of certificate file, as well as the certificate's bytes. In
-// the case of der certificates, it also stores the password.
-#[derive(Clone, Debug, Eq, PartialEq, Serialize, Deserialize)]
-pub(crate) enum CertDetails {
-    Pem(Vec<u8>),
-    Der(Vec<u8>, String),
-}
-
 /// A [Serde][serde]-enabled wrapper around [`reqwest::Identity`].
 #[derive(Clone, Debug, Eq, PartialEq, Serialize, Deserialize)]
 pub struct Identity {
-    pub(crate) cert: CertDetails,
+    der: Vec<u8>,
+    pass: String,
 }
 
 impl Identity {
-    /// Wraps [`reqwest::Identity::from_pem`].
-    pub fn from_pem(pem: &[u8]) -> Result<Self, reqwest::Error> {
-        let _ = reqwest::Identity::from_pem(&pem)?;
-        Ok(Identity {
-            cert: CertDetails::Pem(pem.into()),
-        })
+    /// Reimplements [`reqwest::Identity::from_pem`] in terms of OpenSSL.
+    ///
+    /// The implementation in reqwest requires rustls.
+    pub fn from_pem(pem: &[u8]) -> Result<Self, openssl::error::ErrorStack> {
+        let pkey = PKey::private_key_from_pem(pem)?;
+        let mut certs = Stack::new()?;
+        let mut cert_iter = X509::stack_from_pem(pem)?.into_iter();
+        let cert = cert_iter.next().unwrap();
+        for cert in cert_iter {
+            certs.push(cert)?;
+        }
+        let mut pkcs_builder = Pkcs12::builder();
+        pkcs_builder.ca(certs);
+        let pass = String::new();
+        let friendly_name = "";
```

Comment:
I always get turned around which format has friendly names (and which don't). A comment here would be appreciated, e.g. "pem certificates don't have friendly names, though the pkcs builder requires it" or something.

Reply:
Ah yeah, for sure! Added a comment.